packages/cosmic-swingset/lib/ag-solo/vats/plugin.js

```diff
@@ -0,0 +1,185 @@
+      if (typeof index !== 'number') {
+        // An error string.
+        throw Error(index);
+      }
+
+      // Register our stable callbacks for this connect index.
+      modConnection.init(
+        index,
+        harden({
+          dispatch(obj) {
+            if (obj.epoch !== currentEpoch) {
+              return false;
+            }
+            return currentDispatch(obj);
+          },
+          reset(epoch) {
+            return currentReset(epoch);
+          },
+        }),
+      );
+
+      const connect = () => {
+        // Create a CapTP channel.
+        const myEpoch = nextEpoch;
+        nextEpoch += 1;
+        console.info(
+          `Connecting to ${specifier}.${index} with epoch ${myEpoch}`,
+        );
+        const { getBootstrap, dispatch } = makeCapTP(
+          specifier,
+          obj => {
+            // console.warn('sending', index, obj);
+            D(pluginDevice).send(index, obj);
+          },
+          undefined,
+          { ...vatPowers, epoch: myEpoch },
+        );
+
+        currentReset = _epoch => {
+          bootPK = makePromiseKit();
+
+          // Tell our clients we are resetting.
+          E(resetter).onReset(bootPK.promise.then(_ => true));
+
+          // Attempt to restart the protocol using the same device connection.
+          connect();
+        };
+
+        currentDispatch = obj => {
+          // console.warn('receiving', index, obj);
+          dispatch(obj);
+        };
+
+        currentEpoch = myEpoch;
+
+        // Publish our started plugin.
+        bootPK.resolve(E(getBootstrap()).start(opts));
+      };
+
+      const actions = harden({
+        /**
+         * Create a stable identity that just forwards to the current implementation.
+         */
+        makeStableForwarder(walker = { walk: bootP => bootP }) {
+          let pr;
+          // eslint-disable-next-line no-new
+          new HandledPromise((_resolve, _reject, resolveWithPresence) => {
+            pr = resolveWithPresence({
+              applyMethod(_p, name, args) {
+                // console.warn('applying method epoch', currentEpoch);
+                const targetP = E(walker).walk(bootPK.promise);
+                return HandledPromise.applyMethod(targetP, name, args);
+              },
+              get(_p, name) {
+                // console.warn('applying get epoch', currentEpoch);
+                const targetP = E(walker).walk(bootPK.promise);
+                return HandledPromise.get(targetP, name);
+              },
+            });
+          });
+          return pr;
+        },
+      });
+
+      // Declare the first reset.
+      E(resetter).onReset(false);
+
+      // Start the first connection.
+      connect();
+
+      // Give up our bootstrap object for the caller to use.
+      return harden({
+        // This is the public state, a promise that never resolves,
+        // but pipelines messages to the bootPK.promise.
+        bootstrap: actions.makeStableForwarder(),
```

Comment:
you might name this something other than `bootstrap` to avoid confusion with the first kernel message to the bootstrap vat. "start" or "init" or "prepare" or "setup", combined with "plugin" and/or "manager" or something.

Reply:
It's the CapTP bootstrap object.  I'll call it pluginRoot.